mlperf_logging/rcp_checker/rcp_checker.py

```diff
@@ -0,0 +1,339 @@
+'''
+RCP checker: Verifies convergence points of submissions by comparing them against RCPs (Reference Convergence Points)
+'''
+
+import argparse
+import glob
+import json
+import os
+import numpy as np
+import re
+import scipy.stats
+
+# Number of submission runs for each benchmark
+# References need 2x of these runs
+# We use olympic scoring for statistics, so we reject
+# both the top and bottom reference and submission numbers
+submission_runs = {
+    'bert': 10,
+    'dlrm': 5,
+    'maskrcnn' : 5,
+    'resnet' : 5,
+    'ssd' : 5,
+    'unet3d' : 20,
+    'rnnt': 10,
+}
+
+MLLOG_TOKEN = ':::MLLOG '
+
+
+def get_submission_epochs(result_files):
+    '''
+    Extract convergence epochs from a list of submission files
+    Returns the batch size and the list of epochs to converge
+    -1 means run did not converge. Return None if > 1 files
+    fail to converge
+    '''
+    not_converged = 0
+    subm_epochs = []
+    bs = -1
+    for result_file in result_files:
+        with open(result_file, 'r') as f:
+            file_contents = f.readlines()
+            for line in file_contents:
+                if line.startswith(MLLOG_TOKEN):
+                    str = line[len(MLLOG_TOKEN):]
+                    if "global_batch_size" in str:
+                        # Do we need to make sure global_batch_size is the same
+                        # in all files? If so, this is obviously a bad submission
+                        bs = json.loads(str)["value"]
+                    if "eval_accuracy" in str:
+                        eval_accuracy_str = str
+                    if "run_stop" in str:
+                        # Epochs to converge is the the last epochs value on
+                        # eval_accuracy line before run_stop
+                        conv_result = json.loads(str)["metadata"]["status"]
+                        conv_epoch = json.loads(eval_accuracy_str)["metadata"]["epoch_num"]
+                        if conv_result == "success":
+                            subm_epochs.append(conv_epoch)
+                        else:
+                            subm_epochs.append(-1)
+                            not_converged = not_converged + 1
+    subm_epochs.sort()
+    if not_converged > 1:
+        return None
+    return bs, subm_epochs
+
+
+def eval_submission_record(rcp_record, subm_epochs):
+    '''Compare reference and submission convergence.'''
+    mean_subm_epochs = np.mean(subm_epochs[1:len(subm_epochs)-1])
```

Comment:
Is a precondition for this function that the epochs are sorted? Do we just want to resort here for safety?

Reply:
Good point, I will fix that.